algorithm.md

```diff
@@ -0,0 +1,174 @@
+
+## Edges
+Every relationship between contributors or contributors is represented by an edge in the
+graph. For example, a GitHub pull request may have some of the following edges:
+- an *authored by* edge, connecting it with a user that authored the pull
+- a *has child* edge, connecting it with its comments or reviews
+- a *has parent* edge, connecting it with the repository that contains it
+- a *references* edge, connecting it with an issue or pull it textually references
+
+By convention, we name edges with verb phrases, so that the `src` is the
+subject and the `dst` is the object of the verb phrase. Thus, an edge
+connecting a GitHub user with a post they wrote would be an *authors* edge.
+
+### Edge Bidirectionality
+
+In practice, we often want to flow cred along both directions of an edge. For
+example, if a user authors a post, that user should earn some cred from the
+post (cred flows along the *authored by* edge). However, we may also want cred
+to flow from the user to the post, with the intuition that posts authored by
+important contributors are likely more important than posts authored by someone
+who has no history in the project. Then, cred should flow along the *authors*
+edge.
+
+These situations come up pretty often, so by default, we treat every edge as
+potentially bidirectional. When we assign weights to edges, we give every edge
+a `toWeight` and `froWeight`. The `toWeight` configures how much cred flows
+from the `src` to the `dst`, and the `froWeight` configures how much cred flows
+from the `dst` to the `src`. This way, we don't need to add backwards edges to
+the graph to enable these flows. If an edge really should be uni-directional,
+we can just set one of the weights to 0.
+
+## Plugin Graphs
+
+A core goal for SourceCred is configurability and extensibility. That
+extensibility starts with the graph. The SourceCred core doesn't create any
+graphs itself; it just defines the semantics for graphs. Graphs are actually
+created by plugins. SourceCred currently includes two initial plugins: one for
+Git, another for GitHub. In the future, we imagine having plugins for other
+tools, like Twitter, StackOverflow, Google Docs, npm, and so forth.
+
+### Plugin Namespacing
+
+As mentioned above, we require that node and edge addresses be unique within a
+graph. That means that they need to be unique across plugins. To ensure this
+property, plugins are expected to define their own namespaces within address
+space, by making all of their node and edge addresses start with a unique
+two-part plugin prefix. The first element of the prefix should be the name of
+the owner of the plugin, and the second part should be the name of the plugin
+itself.
+
+For example, every node and edge address created by SourceCred's GitHub plugin
+starts with the prefix `["sourcecred", "github"]`.
+
+## Node and Edge Types
+
+Nodes and edges partition into meaningful "types" or "kinds" which share common
```

Comment:
> Nodes and edges partition into meaningful "types"

Not sure that I quite agree with how this section is formulated.

Is “GitHub node” a type of node? (GitHub nodes have the prefix
`["sourcecred", "github"]`.)

Is “GitHub comment” a type of node? (Comment nodes have the prefix
`["sourcecred", "github", "COMMENT"]`.)

Is “GitHub issue comment” a type of node? (Issue comment nodes have the
prefix `["sourcecred", "github", "COMMENT", "ISSUE"])`.

Intuitively, I’d answer “yes” to all three questions. But then the types
don’t [partition] nodes; we’d need the answer to exactly one of the
above questions to be “yes”. I think that it’s important to use the term
_partition_ precisely, because this has implications when we talk about
(e.g.) [whether the partition assumption is reasonable][p-assum].

[partition]: https://en.wikipedia.org/wiki/Partition_of_a_set
[p-assum]: https://github.com/sourcecred/sourcecred/issues/476#issuecomment-462178059

Reply:
I replaced "partitioned" with "organized".